### PR TITLE
Remove Singleton Configuration

### PR DESCRIPTION
**Pull request by ZacBlanco**

### CODE REVIEW

File: core/common/src/main/java/alluxio/ClientContext.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  private volatile AlluxioConfiguration mConf;
+  private final Subject mSubject;
+
+  /**
+   * A client context with information about the subject and configuration of the client.
+   *
+   * @param subject the security subject to use
+   * @param alluxioConf the {@link AlluxioConfiguration} to use. If null, the site property defaults
+   *     will be loaded
+   * @return A new client context with the specified properties and subject
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/ClientContext.java]: nit: javadoc `@param` and `@return` should start lowercase and not end in period
```
➜  alluxio git:(58f168483e) ✗ git diff upstream/master | grep '@return [A-Z]' | grep '+'
+   * @return A new client context with the specified properties and subject
+   * @return A copy of AlluxioProperties representing the configuration
+   * @return A builder for a {@link MasterClientContext}
+   * @return UFS configuration from a given alluxio configuration
+   * @return The set of Alluxio properties loaded from the site-properties file
+   * @return A new configuration representing the merged properties
+   * @return An configuration object containing the original configuration merged with cluster
+   * @return The proper channel type USER_NETWORK_NETTY_CHANNEL
```

File: core/client/fs/src/main/java/alluxio/client/AbstractOutStream.java [hunk 2]
```diff
@@ -31,7 +31,8 @@
    * @return the number of bytes written to this stream
    * @deprecated this method will not work if more than MAX_INT bytes are written; to get a count of
    *             bytes written, wrap this stream in a counting output stream such as
-   *             org.apache.commons.io.output.CountingOutputStream
+   *             {@link org.apache.commons.io.output.CountingOutputStream}
+   *
```
Comment (aaudiber) [inline core/client/fs/src/main/java/alluxio/client/AbstractOutStream.java]: nit: extra blank line

File: core/common/src/main/java/alluxio/util/ConfigurationUtils.java [hunk 3]
```diff
@@ -247,4 +271,160 @@ public static boolean masterHostConfigured() {
...
+        }
+      }
+    }
+    return sDefaultProperties.copy();
+  }
+
+  /**
+   * Reloads site properties from disk.
+   */
+  public static synchronized void reloadProperties() {
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/util/ConfigurationUtils.java]: Do we need both `synchronized(this)` and `synchronized(DEFAULT_PROPERTIES_LOCK)` or can we use just one?

File: core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 4]
```diff
@@ -461,81 +459,81 @@ public Path getWorkingDirectory() {
...
-    Authority auth = Authority.fromString(uri.getAuthority());
-    if (auth instanceof UnknownAuthority) {
-      // TODO(zac): In Alluxio 2.0 this warning will be upgraded to an exception
-      LOG.warn("Authority \"{}\" is unknown. The client will not be configured with this"
-          + " authority. The authority connection details will be loaded from your client"
-          + " configuration.",
-          auth);
-      mAlluxioHeader = getScheme() + ":///";
-    }
+    synchronized (this) {
```
Comment (aaudiber) [inline core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: To minimize impact of the check on the rest of the method, we can do it like this:
```java
private final AtomicBoolean mCalledInitialize = new AtomicBoolean(false);
public void initialize() {
  Preconditions.checkState(mCalledInitialize.compareAndSet(false, true), "Illegally invoked initialize() multiple times");
  // normal implementation
}
```

File: core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 5]
```diff
@@ -461,81 +459,81 @@ public Path getWorkingDirectory() {
...
-   * Hadoop compatible API and native Alluxio API in the same process.
+   * Sets up a lazy connection to Alluxio through mFileSystem and mFsContext. This must be
+   * called before client operations in order to guarantee the integrity of the
+   * {@link FileSystemContext}.
    *
-   * If hadoop file system cache is enabled, this method should only be called when switching user.
+   * If hadoop file system cache is enabled, this method should only be called when switching
+   * user. If it is called twice on the same object an exception will be thrown.
    */
   @SuppressFBWarnings("ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD")
```
Comment (aaudiber) [inline core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: can this suppression be removed now?

File: core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 6]
```diff
@@ -461,81 +459,81 @@ public Path getWorkingDirectory() {
...
-   * Sets up a lazy connection to Alluxio through mFileSystem. This method will override and
-   * invalidate the current contexts. This must be called before client operations in order to
-   * guarantee the integrity of the contexts, meaning users should not alternate between using the
-   * Hadoop compatible API and native Alluxio API in the same process.
+   * Sets up a lazy connection to Alluxio through mFileSystem and mFsContext. This must be
+   * called before client operations in order to guarantee the integrity of the
+   * {@link FileSystemContext}.
    *
-   * If hadoop file system cache is enabled, this method should only be called when switching user.
+   * If hadoop file system cache is enabled, this method should only be called when switching
```
Comment (aaudiber) [inline core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: if this can't be called twice, it doesn't make sense to say it should be used to switch users

File: core/common/src/main/java/alluxio/ClientContext.java (hunk 1 above)
Comment (ZacBlanco) [reply core/common/src/main/java/alluxio/ClientContext.java]: The periods are caught by checkstyle. If we really want to enforce starting in lower case it should be added to checkstyle. I'll fix these though

Comment (aaudiber) [reply core/common/src/main/java/alluxio/ClientContext.java]: agreed, we can add it after the major outstanding PRs get in

File: core/common/src/main/java/alluxio/util/ConfigurationUtils.java (hunk 3 above)
Comment (ZacBlanco) [reply core/common/src/main/java/alluxio/util/ConfigurationUtils.java]: Well we don't want to reload the properties at the same time because it's a separate public method, but I would still use the `DEFAULT_PROPERTIES_LOCK` because it stops us from reloading the properties unnecessarily. Without the `synchronize(DEFAULT_PROPERTIES_LOCK)` it's possible for another method to attempt to reload the properties before the reference to the `sDefaultProperties` is set

File: core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java (hunk 4 above)
Comment (ZacBlanco) [reply core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: fixed

File: core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java (hunk 5 above)
Comment (ZacBlanco) [reply core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: I'll see if findbugs still produces a warning

File: core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java (hunk 6 above)
Comment (ZacBlanco) [reply core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: sg

File: core/common/src/main/java/alluxio/util/ConfigurationUtils.java (hunk 3 above)
Comment (aaudiber) [reply core/common/src/main/java/alluxio/util/ConfigurationUtils.java]: Can we synchronize on `DEFAULT_PROPERTIES_LOCK` here instead of synchronizing on `this`?

Comment (ZacBlanco) [reply core/common/src/main/java/alluxio/util/ConfigurationUtils.java]: sure

File: core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 7]
```diff
@@ -486,56 +485,52 @@ public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throw
...
-        }
-      } else {
-        initializeInternal(uriConfProperties, conf);
-      }
-      // Must happen inside the lock so that the global filesystem context isn't changed by a
-      // concurrent call to initialize.
-      updateFileSystemAndContext();
-      sInitialized = true;
+    AlluxioProperties alluxioProps = ConfigurationUtils.defaults();
+    AlluxioConfiguration alluxioConf = mergeConfigurations(uriConfProperties, conf, alluxioProps);
```
Comment (gpang) [inline core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: NIT: can we just inline `ConfigurationUtils.defaults()`, since we don't need to reuse that object again?

Comment (ZacBlanco) [reply core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: done

File: core/client/fs/src/main/java/alluxio/client/file/FileSystem.java [hunk 8]
```diff
@@ -71,10 +72,11 @@
     private Factory() {} // prevent instantiation
 
     public static FileSystem get(AlluxioConfiguration alluxioConf) {
-      return get(FileSystemContext.create(alluxioConf));
+      return get(ClientContext.create(alluxioConf));
     }
 
-    public static FileSystem get(FileSystemContext context) {
+    public static FileSystem get(ClientContext context) {
+      Preconditions.checkNotNull(context);
```
Comment (aaudiber) [inline core/client/fs/src/main/java/alluxio/client/file/FileSystem.java]: when using `Preconditions.checkNotNull`, include the name of the variable so that if the check fails, it can help us find where the issue occurred.
```java
Preconditions.checkNotNull(context, "context");
```

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 9]
```diff
@@ -120,14 +126,37 @@
   private final ClientContext mClientContext;
   private final String mAppId;
 
+  /**
+   * Returns a filesystem context which is stored in the cache or creates one and then caches it
+   * for future use.
+   *
+   * @param context The {@link ClientContext} to use
+   * @return A corresponding {@link FileSystemContext}
```
Comment (aaudiber) [inline core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java]: A -> a

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 10]
```diff
@@ -120,14 +126,37 @@
   private final ClientContext mClientContext;
   private final String mAppId;
 
+  /**
+   * Returns a filesystem context which is stored in the cache or creates one and then caches it
+   * for future use.
+   *
+   * @param context The {@link ClientContext} to use
```
Comment (aaudiber) [inline core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java]: The -> the

File: core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 11]
```diff
@@ -504,13 +504,17 @@ public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throw
...
+    // Grab a reference to a FileSystemContext which matches our current configuration and
+    // security subject.
+    mFsContext = FileSystemContext.getOrCreate(ctx);
 
     // Sets the file system.
     //
     // Must happen inside the lock so that the filesystem context isn't changed by a
     // concurrent call to initialize.
-    mFileSystem = FileSystem.Factory.get(mFsContext);
+    mFileSystem = FileSystem.Factory.get(ctx);
```
Comment (aaudiber) [inline core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: is the comment on this out of date?

File: core/common/src/main/java/alluxio/ClientContext.java [hunk 12]
```diff
@@ -118,4 +119,34 @@ public AlluxioConfiguration getConf() {
   public Subject getSubject() {
     return mSubject;
   }
+
+  @Override
+  public int hashCode() {
+    return com.google.common.base.Objects.hashCode(mSubject, mConf);
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/ClientContext.java]: `Objects.hash(mSubject, mConf)`

File: core/common/src/main/java/alluxio/ClientContext.java [hunk 13]
```diff
@@ -118,4 +119,34 @@ public AlluxioConfiguration getConf() {
...
+
+  /**
+   * This method will only return true for this class if the object being compared with this was
+   * created using the constructor which accepts a client context.
+   *
+   * Currently, {@link Object#equals(Object)} is not implemented for
+   * {@link InstancedConfiguration} thus the only way for this to return true are if the
+   * references to mConf are the same which is only possible when the object is created using that
+   * constructor.
+   */
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/ClientContext.java]: Why not implement `equals` and `hashCode` for `InstancedConfiguration`?

File: core/common/src/main/java/alluxio/conf/AlluxioProperties.java [hunk 14]
```diff
@@ -57,6 +59,9 @@
   /** Map of property sources. */
   private final ConcurrentHashMap<PropertyKey, Source> mSources = new ConcurrentHashMap<>();
 
+  /** A value representing the running hash code for this set of properties. */
+  private AtomicInteger mPrecomputedHash = new AtomicInteger(0);
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/conf/AlluxioProperties.java]: Can we implement `hashCode`/`equals` on immutable configuration objects instead? Caching the hash code would be much easier there since it doesn't change. Also, are we certain that it's necessary to perform caching? It seems like we would only call `hashCode` a single time when putting it in the cache.

File: core/common/src/main/java/alluxio/conf/InstancedConfiguration.java [hunk 15]
```diff
@@ -509,6 +509,13 @@ private void checkTieredLocality() {
     }
   }
 
+  @Override
+  public int hashCode() {
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/conf/InstancedConfiguration.java]: if we implement `hashCode` we also need to implement `equals`

File: tests/src/test/java/alluxio/client/fs/ImpersonationIntegrationTest.java [hunk 16]
```diff
@@ -233,7 +236,8 @@ public void impersonationHdfsGroupDenied() throws Exception {
   }
 
   private void checkCreateFile(Subject subject, String expectedUser) throws Exception {
-    FileSystemContext context = FileSystemContext.create(subject, ServerConfiguration.global());
+    FileSystemContext context = FileSystemContext.getOrCreate(ClientContext.create(subject,
```
Comment (aaudiber) [inline tests/src/test/java/alluxio/client/fs/ImpersonationIntegrationTest.java]: why is this test changed?

File: core/common/src/main/java/alluxio/conf/AlluxioProperties.java (hunk 14 above)
Comment (ZacBlanco) [reply core/common/src/main/java/alluxio/conf/AlluxioProperties.java]: The only thing that's immutable is the `AlluxioConfiguration` interface. You can't enforce implementations to implement the `hashCode` method. InstancedConfiguration. I chose to implement it on `AlluxioProperties` rather than `InstancedConfiguration` The idea around pre-computing it is so that when a user creates a configuration and needs to retrieve a client that the hashCode would need to be computed for every call to `Filesystem.get`. Microbench creates a huge number clients and we would like to minimize the performance impact of doing the lookup.

File: core/common/src/main/java/alluxio/conf/InstancedConfiguration.java (hunk 15 above)
Comment (ZacBlanco) [reply core/common/src/main/java/alluxio/conf/InstancedConfiguration.java]: still working on it

File: tests/src/test/java/alluxio/client/fs/ImpersonationIntegrationTest.java (hunk 16 above)
Comment (ZacBlanco) [reply tests/src/test/java/alluxio/client/fs/ImpersonationIntegrationTest.java]: Theoretically before this branch this would have been doing something similar with the original context cache. I changed them to `getOrCreate` instead to see if it would still pass. I did this one specifically because there was a compilation error when I removed the constructor with the `(Subject, AlluxioConfiguration)` parameters

File: core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java (hunk 11 above)
Comment (ZacBlanco) [reply core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: Yeah will fix

File: core/common/src/main/java/alluxio/underfs/MultiRangeObjectInputStream.java [hunk 17]
```diff
@@ -34,6 +33,9 @@
   /** Position the current stream was open till (exclusive). */
   protected long mEndPos;
 
+  /** The block size to read with. */
+  protected long mBlockSize = -1L;
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/underfs/MultiRangeObjectInputStream.java]: Why is this necessary?

Comment (ZacBlanco) [reply core/common/src/main/java/alluxio/underfs/MultiRangeObjectInputStream.java]: Previously we used the block size to perform reads. To keep the same behavior we need the `mBlockSize` parameter.

Comment (ZacBlanco) [reply core/common/src/main/java/alluxio/underfs/MultiRangeObjectInputStream.java]: https://github.com/Alluxio/alluxio/blob/master/core/common/src/main/java/alluxio/underfs/MultiRangeObjectInputStream.java#L115-L141

Comment (calvinjia) [reply core/common/src/main/java/alluxio/underfs/MultiRangeObjectInputStream.java]: Got it, thanks. Could we omit the parameter in `createStream` and use the member variable `mBlockSize` in the `end` calculation?

Comment (ZacBlanco) [reply core/common/src/main/java/alluxio/underfs/MultiRangeObjectInputStream.java]: sgtm

File: core/common/src/main/java/alluxio/conf/AlluxioProperties.java [hunk 18]
```diff
@@ -189,6 +201,15 @@ public void forEach(BiConsumer<? super PropertyKey, ? super String> action) {
     }
   }
 
+  /**
+   * Makes a copy of the backing properties and returns them in a new object.
+   *
+   * @return a copy of the current properties
+   */
+  public AlluxioProperties copy() {
+    return new AlluxioProperties(this);
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/conf/AlluxioProperties.java]: Does this need to be threadsafe?

File: core/common/src/main/java/alluxio/util/ConfigurationUtils.java [hunk 19]
```diff
@@ -247,4 +271,162 @@ public static boolean masterHostConfigured() {
...
   }
+
+  /**
+   * Returns an instance of {@link AlluxioConfiguration} with the defaults and values from
+   * alluxio-site properties.
+   *
+   * @return the set of Alluxio properties loaded from the site-properties file
+   */
+  public static AlluxioProperties defaults() {
+    if (sDefaultProperties == null) {
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/util/ConfigurationUtils.java]: `sDefaultProperties` needs to be volatile for this locking scheme to work?

File: core/common/src/main/java/alluxio/util/ConfigurationUtils.java [hunk 20]
```diff
@@ -247,4 +271,162 @@ public static boolean masterHostConfigured() {
...
+   */
+  public static void reloadProperties() {
+    synchronized (DEFAULT_PROPERTIES_LOCK) {
+      // Step1: bootstrap the configuration. This is necessary because we need to resolve alluxio
+      // .home (likely to be in system properties) to locate the conf dir to search for the site
+      // property file.
+      AlluxioProperties properties = new AlluxioProperties();
+      InstancedConfiguration conf = new InstancedConfiguration(properties);
+      properties.merge(System.getProperties(), Source.SYSTEM_PROPERTY);
+      Properties siteProps = null;
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/util/ConfigurationUtils.java]: Do you mind declaring this closer to when it is used?

File: core/common/src/main/java/alluxio/conf/AlluxioProperties.java (hunk 18 above)
Comment (ZacBlanco) [reply core/common/src/main/java/alluxio/conf/AlluxioProperties.java]: The class is annotated as `NotThreadSafe`. If we want it to be thread safe we should analyze all methods in this class and remove the annotation.

Otherwise- the AlluxioProperties object is never used used outside of the AlluxioConfiguration objects. The `AlluxioConfiguration` interface makes the the configuration objects immutable - which means to modify them you'd need to copy the properties anyways which will only happen in one thread.

Comment (calvinjia) [reply core/common/src/main/java/alluxio/conf/AlluxioProperties.java]: Got it, I was worried about the copy not getting a consistent state since there are two `putAll` calls which are not guaranteed to work if the backing map is being modified.
